_drafts/2020-09-17-fixing-an-n1-performance-problem-in-xaf-xpo-with-totally-undocumented-apis.md

```diff
@@ -0,0 +1,948 @@
+                if (storage.TryGetWideDataItem(GetType().FullName, out var store) && store is Dictionary<Guid, int> cache)
+                {
+                    if (cache.TryGetValue(Oid, out var hourSum))
+                    {
+                        return hourSum;
+                    }
+                }
+            }
+            //We never should reach this point
+            //Do the right thing as fallback anyway
+            return OfferItems.Sum(m => m.Hours);
+        }
+    }
+
+    [Association, Aggregated]
+    public XPCollection<FasterOfferItem> OfferItems => GetCollection<FasterOfferItem>(nameof(OfferItems));
+}
+
+public class FasterOfferItem : BaseObject
+{
+    public FasterOfferItem(Session session) : base(session) { }
+
+    private int _Hours;
+    [Persistent("Hours")]
+    public int Hours { get => _Hours; set => SetPropertyValue(nameof(Hours), ref _Hours, value); }
+
+    private FasterOffer _FasterOffer;
+    [Persistent("FasterOffer"), Association]
+    public FasterOffer FasterOffer { get => _FasterOffer; set => SetPropertyValue(nameof(FasterOffer), ref _FasterOffer, value); }
+}
+```
+
+Let's look at the queries generated in the **Client** mode:
+
+```txt
+19.09.20 15:51:57.286 Executing sql 'select N0."Oid",N0."Name",N0."OptimisticLockField",N0."GCRecord" from "dbo"."FasterOffer" N0 where N0."GCRecord" is null'
+19.09.20 15:51:57.287 Result: rowcount = 300, total = 11950, N0.{Oid,Guid} = 4800, N0.{Name,String} = 4750, N0.{OptimisticLockField,Int32} = 1200, N0.{GCRecord,Int32} = 1200
+19.09.20 15:51:57.289 Executing sql 'select N0."Oid",(select sum(N1."Hours") as Res0 from "dbo"."FasterOfferItem" N1 where ((N0."Oid" = N1."FasterOffer") and N1."GCRecord" is null)) from "dbo"."FasterOffer" N0 where N0."GCRecord" is null'
+19.09.20 15:51:57.359 Result: rowcount = 300, total = 6000, N0.{Oid,Guid} = 4800, SubQuery(Sum,N1.{Hours,Int32},Select N1.{Hours,Int32}
+  from "FasterOfferItem" N1
+ where N0.{Oid,Guid} = N1.{FasterOffer,Guid} And N1.{GCRecord,Int32} Is Null) = 1200
+19.09.20 15:51:57.363 Executing sql 'select top 1 count(*) from "dbo"."FasterOffer" N0 where N0."GCRecord" is null'
+19.09.20 15:51:57.363 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+19.09.20 15:51:57.364 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferItem" N0 where N0."GCRecord" is null'
+19.09.20 15:51:57.387 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+
+```
+
+This results in the following performance
+
+1. Sort by `Name`: 0.082 seconds
+2. Sort by `HourSum`: 0.076 seconds
+
+> **Note**: Cause XAF recreates a `Session` object every time it's refreshed, will fetch the second query only once in the livetime of the session.  
+You can control the cache what ever way you like.  
+Beware that you will pay the cost of calculating the aggregates of ALL objects in the `DetailView` context as well (cause there is no natural way to figure out if you are currently displayed in a `DetailView`).  
```

Comment:
```suggestion
Beware that you will pay the cost of calculating the aggregates of ALL objects in the `DetailView` context as well (because there is no natural way to figure out if you are currently displayed in a `DetailView`).  
```